expat/lib/xmlparse.c

```diff
@@ -1954,22 +1954,16 @@ XML_Parse(XML_Parser parser, const char *s, int len, int isFinal) {
                       &parser->m_position);
     nLeftOver = s + len - end;
     if (nLeftOver) {
-      if (parser->m_buffer == NULL
-          || nLeftOver > parser->m_bufferLim - parser->m_buffer) {
-        /* avoid _signed_ integer overflow */
-        char *temp = NULL;
-        const int bytesToAllocate = (int)((unsigned)len * 2U);
-        if (bytesToAllocate > 0) {
-          temp = (char *)REALLOC(parser, parser->m_buffer, bytesToAllocate);
-        }
-        if (temp == NULL) {
-          parser->m_errorCode = XML_ERROR_NO_MEMORY;
-          parser->m_eventPtr = parser->m_eventEndPtr = NULL;
-          parser->m_processor = errorProcessor;
-          return XML_STATUS_ERROR;
-        }
-        parser->m_buffer = temp;
-        parser->m_bufferLim = parser->m_buffer + bytesToAllocate;
+      // hacky way to avoid ERROR_SUSPENDED from XML_GetBuffer
+      const enum XML_Parsing status = parser->m_parsingStatus.parsing;
+      parser->m_parsingStatus.parsing = XML_PARSING;
+      char *const temp = XML_GetBuffer(parser, len + 1);
+      parser->m_parsingStatus.parsing = status;
+      if (temp == NULL) {
+        // error code was set by XML_GetBuffer
+        parser->m_eventPtr = parser->m_eventEndPtr = NULL;
+        parser->m_processor = errorProcessor;
+        return XML_STATUS_ERROR;
       }
       memcpy(parser->m_buffer, end, nLeftOver);
```

Comment:
Maybe I should change from `parser->m_buffer` to `temp` here, for clarity. And/or add an `assert(parser->m_buffer == temp)`?

Reply:
Right, right... But the code as it stands works because we know from the earlier `if` and `#if` that the internal buffer is empty and `XML_CONTEXT_BYTES` is 0 -- so we don't need to preserve any data.

I think I'll add a comment about these details too; it's far from obvious.